boot/bootutil/src/loader.c

```diff
@@ -773,6 +773,39 @@ boot_validate_slot(struct boot_loader_state *state, int slot,
         goto out;
     }
 
+#if BOOT_IMAGE_NUMBER > 1 && !defined(MCUBOOT_ENC_IMAGES) && defined(MCUBOOT_VERIFY_IMG_ADDRESS)
+    /* Verify that the image in the secondary slot has a reset address
+     * located in the primary slot. This is done to avoid users incorrectly
+     * overwriting an application written to the incorrect slot.
+     */
+    if (area_id == FLASH_AREA_IMAGE_SECONDARY(BOOT_CURR_IMG(state))) {
```

Comment:
I would extract this patch to one function which will be no-operation if configuration is irrelevant, (it might be done like was done for hooks callbacks, using an appropriate macro). This will help with legibility of this quite complicated code.
sth like:
```C
static int ensure_startup_address_or_delete(...)
```